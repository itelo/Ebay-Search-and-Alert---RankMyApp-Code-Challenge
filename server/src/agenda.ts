import { MONGODB_URI } from "./util/secrets";
import Agenda from "agenda";

const agenda = new Agenda({ db: { address: MONGODB_URI } });

export default agenda;
