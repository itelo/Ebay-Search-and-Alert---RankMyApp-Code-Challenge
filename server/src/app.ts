import express from "express";
import compression from "compression"; // compresses requests
import session from "express-session";
import bodyParser from "body-parser";
import lusca from "lusca";
import mongo from "connect-mongo";
import cons from "consolidate";
import path from "path";
import mongoose from "mongoose";

import bluebird from "bluebird";
import { MONGODB_URI, SESSION_SECRET } from "./util/secrets";
import agenda from "./agenda";

const MongoStore = mongo(session);

// Controllers (route handlers)
import * as apiController from "./controllers/api";

// Create Express server
const app = express();

// Connect to MongoDB
const mongoUrl = MONGODB_URI;
(<any>mongoose).Promise = bluebird;

mongoose
  .connect(mongoUrl, { useNewUrlParser: true })
  .then(() => {
    /** ready to use. The `mongoose.connect()` promise resolves to undefined. */
  })
  .catch(err => {
    console.log(
      "MongoDB connection error. Please make sure MongoDB is running. " + err
    );
    // process.exit();
  });

agenda.start();

// Express configuration
app.set("port", process.env.PORT || 3000);
app.set("views", path.join(__dirname, "../public"));
app.engine("html", cons.swig);

app.use(compression());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));
app.use(
  session({
    resave: true,
    saveUninitialized: true,
    secret: SESSION_SECRET,
    store: new MongoStore({
      url: mongoUrl,
      autoReconnect: true
    })
  })
);

app.use(lusca.xframe("SAMEORIGIN"));
app.use(lusca.xssProtection(true));
app.use((req, res, next) => {
  res.locals.user = req.user;
  next();
});
app.use((req, res, next) => {
  // After successful login, redirect back to the intended page
  if (
    !req.user &&
    req.path !== "/login" &&
    req.path !== "/signup" &&
    !req.path.match(/^\/auth/) &&
    !req.path.match(/\./)
  ) {
    req.session.returnTo = req.path;
  } else if (req.user && req.path == "/account") {
    req.session.returnTo = req.path;
  }
  next();
});

app.use(
  express.static(path.join(__dirname, "../public"), { maxAge: 31557600000 })
);

/**
 * API examples routes.
 */
app.get("/api/query", apiController.query);
app.post("/api/register", apiController.register);
app.delete("/api/cancel/:_id", apiController.cancel);
app.get("/api/list", apiController.list);

app.get("/*", (req, res) => {
  res.render("index.html");
});

export default app;
