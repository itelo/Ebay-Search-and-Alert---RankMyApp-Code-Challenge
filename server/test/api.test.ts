import request from "supertest";
import app from "../src/app";
import agenda from "../src/agenda";

describe("GET /api", () => {
  describe("get /api/query", () => {
    it("should return 200 OK", () => {
      return request(app)
        .get("/api/query?searchPhrase=iphone")
        .expect(200);
    });
  });

  describe("post /register", () => {
    beforeEach(async done => {
      const jobs = await agenda.jobs({});
      jobs.forEach(async job => {
        await job.remove();
      });

      done();
    });

    it("should return 200 success true", done => {
      return request(app)
        .post("/api/register")
        .send({
          searchPhrase: "pokemon",
          interval: "10",
          email: "someemail@gmail.com"
        })
        .expect(200)
        .end((err, res) => {
          expect(res.body.success).toBe(true);

          expect(res.body.data).toHaveProperty("_id");
          expect(res.body.data).toHaveProperty("email", "someemail@gmail.com");
          expect(res.body.data).toHaveProperty("searchPhrase", "pokemon");
          expect(res.body.data).toHaveProperty("interval", "10 minutes");

          const { _id, ...data } = res.body.data;
          const { success } = res.body;
          const toSnap = {
            data,
            success
          };

          expect(toSnap).toMatchSnapshot();

          if (err) return done(err);
          done();
        });
    });

    it("should return 200 success false", done => {
      return request(app)
        .post("/api/register")
        .send({
          searchPhrase: "pokemon",
          interval: "10",
          email: "someemail@gmail.com"
        })
        .expect(200)
        .end((e1, r1) => {
          return request(app)
            .post("/api/register")
            .send({
              searchPhrase: "pokemon",
              interval: "10",
              email: "someemail@gmail.com"
            })
            .expect(200)
            .end((err, res) => {
              expect(res.body.success).toBe(false);
              expect(res.body.data).toHaveProperty(
                "message",
                "search phrase already registered"
              );

              expect(res.body).toMatchSnapshot();

              if (err) return done(err);
              done();
            });
        });
    });

    it("should return 200 success true", done => {
      return request(app)
        .post("/api/register")
        .send({
          searchPhrase: "pokemon",
          interval: "10",
          email: "someemail@gmail.com"
        })
        .expect(200)
        .end((e1, r1) => {
          request(app)
            .post("/api/register")
            .send({
              searchPhrase: "pokemon",
              interval: "10",
              email: "someemail2@gmail.com"
            })
            .expect(200)
            .end((err, res) => {
              expect(res.body.success).toBe(true);
              expect(res.body.data).toHaveProperty("_id");
              expect(res.body.data).toHaveProperty(
                "email",
                "someemail2@gmail.com"
              );
              expect(res.body.data).toHaveProperty("searchPhrase", "pokemon");
              expect(res.body.data).toHaveProperty("interval", "10 minutes");

              const { _id, ...data } = res.body.data;
              const { success } = res.body;
              const toSnap = {
                data,
                success
              };

              expect(toSnap).toMatchSnapshot();

              if (err) return done(err);
              done();
            });
        });
    });

    it("should return 200 success false due to wrong interval", done => {
      return request(app)
        .post("/api/register")
        .send({
          searchPhrase: "pokemon",
          interval: "99",
          email: "someemail2@gmail.com"
        })
        .expect(200)
        .end((err, res) => {
          expect(res.body.success).toBe(false);
          expect(res.body.data).toHaveProperty("message", "interval not valid");

          expect(res.body).toMatchSnapshot();

          if (err) return done(err);
          done();
        });
    });
  });

  describe("get /list", () => {
    beforeEach(async done => {
      const jobs = await agenda.jobs({});
      jobs.forEach(async job => {
        await job.remove();
      });

      done();
    });

    it("should return 200 success true empty array", done => {
      return request(app)
        .get("/api/list")
        .expect(200)
        .end((err, res) => {
          expect(res.body.success).toBe(true);
          expect(res.body.data).toHaveLength(0);

          expect(res.body).toMatchSnapshot();

          if (err) return done(err);
          done();
        });
    });

    it("should return 200 success true with 1 elem in data", done => {
      return request(app)
        .post("/api/register")
        .send({
          searchPhrase: "pokemon",
          interval: "10",
          email: "someemail@gmail.com"
        })
        .expect(200)
        .end((err, res) => {
          return request(app)
            .get("/api/list")
            .expect(200)
            .end((err, res) => {
              expect(res.body.success).toBe(true);
              expect(res.body.data).toHaveLength(1);

              expect(res.body.data[0]).toHaveProperty("_id");
              expect(res.body.data[0]).toHaveProperty(
                "email",
                "someemail@gmail.com"
              );
              expect(res.body.data[0]).toHaveProperty(
                "searchPhrase",
                "pokemon"
              );
              expect(res.body.data[0]).toHaveProperty("interval", "10 minutes");

              const { _id, ...data } = res.body.data[0];
              const { success } = res.body;
              const toSnap = {
                data: [data],
                success
              };

              expect(toSnap).toMatchSnapshot();

              if (err) return done(err);
              done();
            });
        });
    });
  });

  describe("delete /api/cancel/id", () => {
    beforeEach(async done => {
      const jobs = await agenda.jobs({});
      jobs.forEach(async job => {
        await job.remove();
      });

      done();
    });

    it("should return 200 success true", done => {
      return request(app)
        .post("/api/register")
        .send({
          searchPhrase: "pokemon",
          interval: "10",
          email: "someemail@gmail.com"
        })
        .expect(200)
        .end((err, res) => {
          const idToDelete = res.body.data._id;
          return request(app)
            .delete(`/api/cancel/${idToDelete}`)
            .expect(200)
            .end((err, res) => {
              expect(res.body.success).toBe(true);

              if (err) return done(err);
              done();
            });
        });
    });

    it("should return 200 success false", done => {
      const idToDelete = "20";
      return request(app)
        .delete(`/api/cancel/${idToDelete}`)
        .expect(200)
        .end((err, res) => {
          expect(res.body.success).toBe(false);

          if (err) return done(err);
          done();
        });
    });
  });
});
