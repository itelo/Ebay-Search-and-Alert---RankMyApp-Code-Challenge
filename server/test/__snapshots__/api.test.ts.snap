// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /api get /list should return 200 success true empty array 1`] = `
Object {
  "data": Array [],
  "success": true,
}
`;

exports[`GET /api get /list should return 200 success true with 1 elem in data 1`] = `
Object {
  "data": Array [
    Object {
      "email": "someemail@gmail.com",
      "interval": "10 minutes",
      "searchPhrase": "pokemon",
    },
  ],
  "success": true,
}
`;

exports[`GET /api post /register should return 200 success false 1`] = `
Object {
  "data": Object {
    "message": "search phrase already registered",
  },
  "success": false,
}
`;

exports[`GET /api post /register should return 200 success false due to wrong interval 1`] = `
Object {
  "data": Object {
    "message": "interval not valid",
  },
  "success": false,
}
`;

exports[`GET /api post /register should return 200 success true 1`] = `
Object {
  "data": Object {
    "email": "someemail@gmail.com",
    "interval": "10 minutes",
    "searchPhrase": "pokemon",
  },
  "success": true,
}
`;

exports[`GET /api post /register should return 200 success true 2`] = `
Object {
  "data": Object {
    "email": "someemail2@gmail.com",
    "interval": "10 minutes",
    "searchPhrase": "pokemon",
  },
  "success": true,
}
`;
